src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1361,15 +1434,31 @@ public SolutionCompilationState WithFrozenSourceGeneratedDocuments(
                     // Just compute the checksum from the source text passed in.
                     originalSourceTextChecksum: null,
                     generationDateTime);
-                documentStates.Add(newGeneratedState);
+
+                if (syntaxNode != null)
+                {
+                    newGeneratedState = newGeneratedState.WithSyntaxRoot(syntaxNode);
+                }
+
+                documentStates.Add(newGeneratedState.Id, newGeneratedState);
+            }
+        }
+
+        if (FrozenSourceGeneratedDocumentStates is not null)
+        {
+            // We also carry forward any previously frozen source generated documents that were not part of this freeze. This ensures multiple
+            // freezes are additive, and everything will be unfrozen correctly, without us needing to have a chain of compilation trackers
+            foreach (var (id, state) in FrozenSourceGeneratedDocumentStates.States)
```

Comment:
cool.  i assume we have a test for the comment above?  if not, add.

Reply:
Yep, we do.